### PR TITLE
VACMS-4260 4264 4265 4266 4267 4276 4277 4278 D9 upgrades

### PR DESCRIPTION
**Pull request by ethanteague**
## Description
See #4260 #4264 #4265 #4266 #4267 #4276 #4277 #4278

## Testing done
Build, log checks, behat, phpunit, spot checks

## Screenshots
N/A

## QA steps

- Spot check multiple pages and review logs for issues.

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [x] `Core Application Team`
- [ ] `Product Support Team`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change? 
- [ ] Yes, and it's written in issue ____ and queued for publication. 
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written 
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [ ] No announcement is needed for this code change. 
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

**Review by acrollet — COMMENTED** [review]
@ethanteague nice, thanks for this! Do you know if we have any existing tests targeting the functionality that `hms_field` provides?

**Comment by ethanteague** [comment]
> @ethanteague nice, thanks for this! Do you know if we have any existing tests targeting the functionality that `hms_field` provides?

Don't believe we do, @acrollet . Think we should write one?

**Comment by acrollet** [comment]
> > @ethanteague nice, thanks for this! Do you know if we have any existing tests targeting the functionality that `hms_field` provides?
> 
> Don't believe we do, @acrollet . Think we should write one?

@ethanteague I think it would make sense, I'm assuming that the module directly affects the content experience/integrity so it would be good to have something in place :-)

**Comment by ethanteague** [comment]
> > > @ethanteague nice, thanks for this! Do you know if we have any existing tests targeting the functionality that `hms_field` provides?
> > 
> > 
> > Don't believe we do, @acrollet . Think we should write one?
> 
> @ethanteague I think it would make sense, I'm assuming that the module directly affects the content experience/integrity so it would be good to have something in place :-)

You thinking this should be a cypress test? If we go this way, is it okay to break the test off into a separate task, @acrollet ? Asking because I'm a cypress noob / hoping to squeak this pr in before Friday.

**Comment by acrollet** [comment]
> You thinking this should be a cypress test? If we go this way, is it okay to break the test off into a separate task, @acrollet ? Asking because I'm a cypress noob / hoping to squeak this pr in before Friday.

I'm agnostic on how we test, I'd just like to identify the use case that we installed the module for and make sure there's some kind of test coverage on it. :-) There are a lot of updates in this PR and the `cer` and `hms_field` modules stick out to me as being somewhat higher-risk.

**Comment by ethanteague** [comment]
> > You thinking this should be a cypress test? If we go this way, is it okay to break the test off into a separate task, @acrollet ? Asking because I'm a cypress noob / hoping to squeak this pr in before Friday.
> 
> I'm agnostic on how we test, I'd just like to identify the use case that we installed the module for and make sure there's some kind of test coverage on it. :-) There are a lot of updates in this PR and the `cer` and `hms_field` modules stick out to me as being somewhat higher-risk.

:+1: test added.

**Review by acrollet — APPROVED** [review]
Thanks for the new tests @ethanteague !

**Review by swirtSJW — COMMENTED** [review]
Thanks for all the work on this.  

I am a little hesitant about pinning all these modules to the dev branch.  I understand why you had to do it.... but I don't like that we have to go that route as there is a bit of risk of having these float with dev branch commits.   I want to raise the issue with @indytechcook  just for him to weigh in on it.

**Comment by indytechcook** [comment]
> I am a little hesitant about pinning all these modules to the dev branch. I understand why you had to do it.... but I don't like that we have to go that route as there is a bit of risk of having these float with dev branch commits. I want to raise the issue with @indytechcook just for him to weigh in on it.

Good catch @swirtSJW try to pin to a commit if you need to use a dev release.  That way we can make sure the patches always apply.  See site_alert for an example: `"drupal/site_alert": "1.x-dev#d3700174bdddeca1f33c6995fde86e436b034270",`

File: composer.json [hunk 1]
```diff
@@ -250,6 +251,10 @@
         "composer-exit-on-patch-failure": true,
         "patches-ignore": {},
         "patches": {
+            "drupal/cer": {
+                "Make d9 compatible": "https://www.drupal.org/files/issues/2021-02-11/3197953.patch",
```
Comment (indytechcook) [inline composer.json:255]: Please update the names used in the patches to include the d.o issue.

example:
`"3197953  - Make d9 compatible": "https://www.drupal.org/files/issues/2021-02-11/3197953.patch",`

Comment (ethanteague) [reply composer.json:255]: Good call / catch, @indytechcook :+1:

**Comment by ethanteague** [comment]
> > I am a little hesitant about pinning all these modules to the dev branch. I understand why you had to do it.... but I don't like that we have to go that route as there is a bit of risk of having these float with dev branch commits. I want to raise the issue with @indytechcook just for him to weigh in on it.
> 
> Good catch @swirtSJW try to pin to a commit if you need to use a dev release. That way we can make sure the patches always apply. See site_alert for an example: `"drupal/site_alert": "1.x-dev#d3700174bdddeca1f33c6995fde86e436b034270",`

Glad you flagged this, @swirtSJW ! @indytechcook - great tip - will do :+1:

**Review by swirtSJW — APPROVED** [review]
Nice work Ethan.

**Comment by swirtSJW** [comment]
image

**Comment by swirtSJW** [comment]
The addition of csv_serialization just broke out csv migrations by altering the version of League CSV.   I am looking into a fix now .... if I can't come up with one tonight we may have to roll this back before tomorrow's release.

**Pull request merged**